### ISSUE TITLE
When DateAdded is blank in one or more Rekordbox tracks, conversion fails

### ISSUE THREAD

**Issue by craigspree** [digital-dj-tools/dj-data-converter#27]
Hi,

I'm using build 0.4.0 on Windows 10. I am able to convert a collection.nml file to a Rekordbox.xml file using the app via command line. However, I can't convert a Rekordbox.xml file to a collection.nml file.

The Rekordbox.xml file was exported from RekordBox 5.7.0. When I attempt to convert it to Traktor, I get an error after 30 seconds saying to check the error-report.edn. A collection.nml file is never created.

Here is the text of the error message in my error-report.edn

`{:args {:input-file "rekordbox.xml"},
 :opts {},
 :error
 {:via
  [{:type #object[E],
    :message
    "Text '2013-3-1' could not be parsed at index 5: 2013-3-1, at index: 5",
    :at
    "DateTimeParseException: Text '2013-3-1' could not be parsed at index 5: 2013-3-1, at index: 5"}],
  :trace
  ["DateTimeParseException: Text '2013-3-1' could not be parsed at index 5: 2013-3-1, at index: 5"
   "at DateTimeFormatter._parseToBuilder (C:\\snapshot\\s\\node_modules\\js-joda\\dist\\js-joda.js:6271:27)"
   "at DateTimeFormatter.parse2 (C:\\snapshot\\s\\node_modules\\js-joda\\dist\\js-joda.js:6239:36)"
   "at DateTimeFormatter.parse (C:\\snapshot\\s\\node_modules\\js-joda\\dist\\js-joda.js:6218:29)"
   "at Function.parse (C:\\snapshot\\s\\node_modules\\js-joda\\dist\\js-joda.js:8509:30)"
   "at Function.cljc.java_time.local_date.parse.cljs$core$IFn$_invoke$arity$2 (C:\\snapshot\\s\\out\\cli.js:0:0)"
   "at cljc.java_time.local_date.parse (C:\\snapshot\\s\\out\\cli.js:0:0)"
   "at converter.time.string__GT_date (C:\\snapshot\\s\\out\\cli.js:0:0)"
   "at f (C:\\snapshot\\s\\out\\cli.js:0:0)"
   "at c (C:\\snapshot\\s\\out\\cli.js:0:0)"
   "at spec_tools.core.Spec.cljs$spec$alpha$Spec$conform_STAR_$arity$2 (C:\\snapshot\\s\\out\\cli.js:0:0)"]}}
`

Attached is a zip with both the rekordbox.xml and the error-report.edn.

Thank you for the awesome tool! It works great for Traktor -> Rekordbox, and I'm equally excited to be able to transfer cues from Rekordbox -> Traktor!
[error-report.zip](https://github.com/digital-dj-tools/dj-data-converter/files/3776530/error-report.zip)

**Comment by alzadude** [comment]
Hi @craigspree thanks for the report.

I see three options:

1. Filter out these items (i.e. keep date added required)
2. Allow these items unchanged
3. Set date added to "today" for these items.

Thoughts?

**Comment by craigspree** [comment]
Hi @alzadude thank you again for an awesome, and vital, app! All of your options are reasonable - is there any difference in effort between them?

I think 2 is likely the best so it doesn't throw off anyone's meta data, if they are all of equal effort.

Thank you, again!

**Comment by alzadude** [comment]
> Hi @alzadude thank you again for an awesome, and vital, app! All of your options are reasonable - is there any difference in effort between them?
> 
> I think 2 is likely the best so it doesn't throw off anyone's meta data, if they are all of equal effort.
> 
> Thank you, again!
> 
> PS: I want to donate, and I see on the main distro page that donations are welcome, but I can't figure out how to donate to you. Please tell me how to go about donating because I am obviously too dense to figure it out on my own :)

Hi @craigspree thanks for the feedback, I'm trying to decide between 2 and 3 🙂

Re: donations, I've recently been accepted into the GitHub Sponsors program, so you can now sponsor me [here](https://github.com/sponsors/alzadude). Thanks!

**Comment by craigspree** [comment]
I think #2 is probably the best if you can't pick between them. I've also just sponsored you. Thanks!

**Comment by alzadude** [comment]
Amazing @craigspree thanks so much! You're my first sponsor 🎉👏🙂

Ok I agree on going for option 2, I'll let you know when it's ready to test!

Thanks again 🙂

**Comment by alzadude** [comment]
Hi @craigspree,

I've built a version of the app with a candidate fix for this issue: https://dev.azure.com/digital-dj-tools/dj-data-converter/_build/results?buildId=226&view=artifacts&type=publishedArtifacts

If you get a moment, would you be able to download and try it out, to verify that the issue is fixed for you?

Thanks!

Alex.

**Comment by craigspree** [comment]
Hi @alzadude Thank you - I was busy with the holidays but just got a chance to test this out. Unfortunately I'm getting an error. Attached is the error report.
[error-report.zip](https://github.com/digital-dj-tools/dj-data-converter/files/4050069/error-report.zip)

**Comment by alzadude** [comment]
Hi @craigspree,

I've just had a look at your `error-report.edn` the error is:
```
Text '2013-3-1' could not be parsed at index 5: 2013-3-1, at index: 5
```
So it looks like we're past the blank DateAdded problem, which is good because it indicates the fix is working 👍 but we're now hitting a different problem.. 👎

I assume date formats in the Rekordbox xml are ISO-8601, which in this case would be `2013-03-01`. `2013-3-1` doesn't have the padding, and so the parsing fails.

Coincidentally, Traktor actually uses the non-padded format, and since version [0.4.0](https://github.com/digital-dj-tools/dj-data-converter/releases/tag/0.4.0) I convert this to ISO-8601 when converting from Traktor to Rekordbox, because I observed that Rekordbox has issues with the non-padded format (see #26).

I'm wondering how non-padded dates got into your Rekordbox data.. I suspect it might be because you created the Rekordbox xml with an earlier version of the app?

When I was working on the fix for #26 myself, unfortunately I found that re-creating the Rekordbox xml with the correct date format and re-importing it did not fix the problem. My hypothesis was that the previous xml file with the incorrect date format was accepted by Rekordbox and was therefore in the Rekordbox database already, and Rekordbox was not able to replace the incorrect date format data with the correct data. Personally I view this lack of date validation as a bug in Rekordbox 👎 which I have passed on to Pioneer DJ..

In the meantime, if you did use an earlier version of the app, the solution will be to repeat the same steps as me:

1. Uninstall Rekordbox 
2. Backup and then delete the directory containing the Rekordbox database, `C:\Users\youruser name\AppData\Roaming\Pioneer\rekordbox` on Windows and `/yourharddrivename/Users/yourusername/Library/Pioneer/rekordbox` on a Mac; if your goal is to convert all your data from Traktor anyway, deleting this shouldn't represent a problem, because the data will be re-created at step 6
3. Reinstall Rekordbox 
4. Re-create the Rekordbox xml using version 0.4.0.
5. Do a quick search in the Rekordbox xml for `2013-3-1`; if everything is working as expected, you shouldn't get any results, and you can proceed to step 6; if you do get a match for unpadded dates, please let me know!
6. Import this xml file with correct date format into Rekordbox.

Note: you might be able to skip steps 1 and 3, but I didn't test that yet!

Thanks,

Alex.

**Comment by alzadude** [comment]
Hi @craigspree, since the DateAdded blank problem is fixed, I've decided to merge this change and make a new release for it (0.4.1).

Regarding the separate date parsing problem you then hit afterwards, please let me know if the steps above resolved the issue. Thanks 🙂
